src/Symfony/Component/JsonStreamer/ValueTransformer/ScalarToValueObjectValueTransformer.php

```diff
@@ -11,30 +11,59 @@
 
 namespace Symfony\Component\JsonStreamer\ValueTransformer;
 
+use BcMath\Number;
 use Symfony\Component\JsonStreamer\Exception\InvalidArgumentException;
 use Symfony\Component\TypeInfo\Type;
-use Symfony\Component\TypeInfo\Type\BuiltinType;
-use Symfony\Component\TypeInfo\TypeIdentifier;
 
 /**
- * Transforms string to DateTimeImmutable during stream reading.
+ * Transforms a scalar to its related value object.
+ *
+ * Does nothing if the value type is not valid.
  *
  * @author Mathias Arlaud <mathias.arlaud@gmail.com>
  *
  * @experimental
  */
-final class StringToDateTimeValueTransformer implements ValueTransformerInterface
+final class ScalarToValueObjectValueTransformer implements ValueTransformerInterface
 {
-    public const FORMAT_KEY = 'date_time_format';
+    public const DATE_TIME_FORMAT_KEY = 'date_time_format';
+
+    /**
+     * @param class-string $valueObjectClassName
+     */
+    public function __construct(
+        private string $valueObjectClassName,
+    ) {
+    }
 
-    public function transform(mixed $value, array $options = []): \DateTimeImmutable
+    public function transform(mixed $value, array $options = []): mixed
     {
-        if (!\is_string($value) || '' === trim($value)) {
-            throw new InvalidArgumentException('The JSON value is either not an string, or an empty string, or null; you should pass a string that can be parsed with the passed format or a valid DateTime string.');
+        if (\DateTimeInterface::class === $this->valueObjectClassName) {
```

Comment:
This logic being totally different for each class is a reason to have separate value transformer classes instead of having a single value transformer implementation that takes a class-string as constructor argument but actually supports only a few class names.
The current proposal makes the code confusing, but more importantly, it does not define a clear contract (which is then a pain when having to maintain BC).